### PR TITLE
Add Horse Animation API

### PR DESCRIPTION
**Pull request by Owen1212055**
This allows you to set things like if a horse is on its hind legs, eating, etc. 

Feedback on a better method name for ``isEating`` and ``setEating`` appreciated. But it causes the horse to sort of "bob" its head. It can be seen normally by feeding a horse wheat,

### CODE REVIEW

File: Spigot-Server-Patches/0696-Horse-Api.patch
Comment (Prof-Bloodstone) [inline Spigot-Server-Patches/0696-Horse-Api.patch]: Obfhelpers should go right above method they are for, not below. We want to conflict when its name or signature changes.
Applies to all other obfhelpers.

File: Spigot-Server-Patches/0696-Horse-Api.patch
Comment (Prof-Bloodstone) [inline Spigot-Server-Patches/0696-Horse-Api.patch]: Not sure I'd call these obfhelpers, but I also didn't see how the team handles such additions normally.

**Comment by Owen1212055** [comment]
I added newlines infront of paper start (according to https://github.com/PaperMC/Paper/pull/5598/files)? Unless I am misunderstanding.

I also got rid of the helpers that weren't actually needed. However, this makes me directly call get/set horse flag in CraftAbstractHorse which I would like feedback on.

**Comment by electronicboy** [comment]
magic values need diff busters (i.e. a comment or something to bust the merge if those values change)

**Comment by Owen1212055** [comment]
> magic values need diff busters (i.e. a comment or something to bust the merge if those values change)

“Bust” the merge? Sorry, but I don’t really understand what you mean here. What should I put in this comment?

File: Spigot-Server-Patches/0696-Horse-Api.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0696-Horse-Api.patch]: Modify the lines that specify these integers to use static fields with them instead so we can reuse here.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0696-Horse-Api.patch
Comment (Proximyst) [inline Spigot-Server-Patches/0696-Horse-Api.patch]: Obfhelpers go on top.

**Review by Proximyst — APPROVED** [review]
Not precisely what I meant, but this looks great nonetheless. Good job and thanks for the contribution!

**Comment by Owen1212055** [comment]
image
Should I instead call setHorseFlag/getHorseFlag in the craftabstracthorse class instead for these methods?

**Comment by kennytv** [comment]
Can you update this to 1.17.1? All the previous issues should be gone as well with obfhelpers no longer being required here. 😛

**Comment by Owen1212055** [comment]
Huh, I guess I forgot about this one when updating my pull requests. Going to merge this into my generic "entity behavior patch".

**Comment by Owen1212055** [comment]
Okay, now this has been updated. As a result however, this now requires https://github.com/PaperMC/Paper/pull/5744.

**Review by Machine-Maker — APPROVED** [review]
Looks fine. Another note that this is blocked by #5744

**Comment by Owen1212055** [comment]
Rebased and ready! No longer blocked

**Comment by jpenilla** [comment]
Setting rearing to false doesn't seem to effectively stop the animation after previously having set it to true. If you look in the `stand` method, you can see vanilla sets the `standCounter` which in the tick method is used to control the standing animation. Maybe it's better to have a `stand()` or `rear()` method which calls `stand()`? Either that, or setting it to false should be fixed.

**Comment by Owen1212055** [comment]
> Setting rearing to false doesn't seem to effectively stop the animation after previously having set it to true. If you look in the `stand` method, you can see vanilla sets the `standCounter` which in the tick method is used to control the standing animation. Maybe it's better to have a `stand()` or `rear()` method which calls `stand()`? Either that, or setting it to false should be fixed.

 I can't confirm this, testing with a horse my current code seems to work fine. Can you show me what you did?
```java
 if (entity instanceof AbstractHorse horse) {
       horse.setRearing(!horse.isRearing());
 }
```

**Comment by jpenilla** [comment]
I ran the same test I did before and now I'm getting different results than before... not sure what happened there. Rearing and eating both seem to work mostly fine, as does eating grass, however eating grass will disable itself 50 ticks after you enable it due to logic hardcoded in the aiStep method of AbstractHorse.

That's really my main reservation here compared to the cat methods. In the example of the cat, the methods we exposed are only called from goals, so by removing existing goals and adding your own the behavior could be controlled predictably. In the case of the horse however, the hardcoded logic in the tick method is going to be messing with these even if you remove all goals, which might be confusing when trying to use the API.

I don't really see a way around this, and I don't think it should necessarily be a merge blocker, just something to think about in how this would be slightly more caveated than some of the other additional behavior APIs.

File: patches/server/0712-Missing-Entity-Behavior-API.patch
Comment (jpenilla) [inline patches/server/0712-Missing-Entity-Behavior-API.patch]: please use explicit `this.` for accessing instance fields/methods

File: patches/server/0712-Missing-Entity-Behavior-API.patch
Comment (jpenilla) [inline patches/server/0712-Missing-Entity-Behavior-API.patch]: personally I would just add public methods `setStanding(boolean)`, `isMouthOpen()`, and `setMouthOpen(boolean)` in `net.minecraft.world.entity.animal.hosrse.AbstractHorse` and call those here instead of using ATs, but I guess either way is fine.

File: build-data/paper.at
Comment (jpenilla) [inline build-data/paper.at]: If you decide to keep the ATs, please remove this comment as well as the line breaks between entries for the same patch. Generally we add a comment with the patch name and list out the ATs below. (Meaning `# Missing Entity API` should also be changed to `# Missing Entity Behavior API`)

**Review by jpenilla — CHANGES_REQUESTED** [review]

**Comment by Owen1212055** [comment]
> I ran the same test I did before and now I'm getting different results than before... not sure what happened there. Rearing and eating both seem to work mostly fine, as does eating grass, however eating grass will disable itself 50 ticks after you enable it due to logic hardcoded in the aiStep method of AbstractHorse.
> 
> 
> 
> That's really my main reservation here compared to the cat methods. In the example of the cat, the methods we exposed are only called from goals, so by removing existing goals and adding your own the behavior could be controlled predictably. In the case of the horse however, the hardcoded logic in the tick method is going to be messing with these even if you remove all goals, which might be confusing when trying to use the API.
> 
> 
> 
> I don't really see a way around this, and I don't think it should necessarily be a merge blocker, just something to think about in how this would be slightly more caveated than some of the other additional behavior APIs.

Yea so, it may be optimal to add some kind of method but I'm not sure. Currently methods like in the fox setsleeping currently resets each tick. I get around this by manually setting it in a mob goal each tick. However, we could add some kind of override that'll cause the tick method to not change that or we could just include in the javadocs it automatically stops.

**Comment by jpenilla** [comment]
Given there's already methods with similar caveats this is probably fine, maybe add something to the docs (for this new method and the aforementioned fox one) if you feel it's appropriate, not sure myself, as it's kind of an implementation detail (maybe someone else has some input?).

Like I said I don't think it should be a merge blocker, so this would get my approval after requested changes are addressed.

**Pull request merged**
